### PR TITLE
Check nested weak types in intersections on target side of relation

### LINKED ISSUE #51043: Combination of intersections/partials causes invalid nested object type to not report error
# Bug Report

### 🔎 Search Terms

partial nested object type

### 🕗 Version & Regression Information

- This changed between versions 4.5.5 and 4.6.4. It reproduces in all versions since 4.6.4 (including nightly)

### ⏯ Playground Link

[Playground link with relevant code](https://tsplay.dev/wRl4Ym)

### 💻 Code

```ts
type OverridesInput = {
    someProp?: 'A' | 'B'
}

const foo: Partial<{something: any}> & {variables: {
    overrides?: OverridesInput;
} & Partial<{
    overrides?: OverridesInput;
}>} = {variables: {overrides: false}};
```

### 🙁 Actual behavior

There is no error on the `overrides` field in the object. Changing the type of `foo` slightly (removing one of the intersections) causes the error to appear correctly, but this unique combination of types produces no error. Does not matter what the contents of `OverridesInput` is, nor what the content inside the outer partial (currently `{something: any}`) is. Additionally, the following all produce the error correctly:

```ts
type OverridesInput = {
    someProp?: 'A' | 'B'
}

const foo: {variables: {
    overrides?: OverridesInput;
} & Partial<{
    overrides?: OverridesInput;
}>} = {variables: {overrides: false}};
```

```ts
type OverridesInput = {
    someProp?: 'A' | 'B'
}

const foo: Partial<{something: any}> & {variables: {
    overrides?: OverridesInput;
}} = {variables: {overrides: false}};
```

```ts
type OverridesInput = {
    someProp?: 'A' | 'B'
}

const foo: Partial<{something: any}> & {variables: Partial<{
    overrides?: OverridesInput;
}>} = {variables: {overrides: false}};
```

### 🙂 Expected behavior

There should be an error on the `overrides` field of the object since `false` is not assignable to `OverridesInput | undefined`.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #51043.

This PR also moves some intersection property check logic from `isRelatedTo` to `structuredTypeRelatedTo` (i.e. from the un-cached to the cached side of relationship checking). This allows us to get rid of a separate cache lookup and two associated `IntersectionState` flags.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -19112,7 +19110,7 @@ namespace ts {
                         }
                     }
 
-                    const isPerformingCommonPropertyChecks = (relation !== comparableRelation || !(source.flags & TypeFlags.Union) && isLiteralType(source)) &&
+                    const isPerformingCommonPropertyChecks = (relation !== comparableRelation || isUnitType(source)) &&
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:19113]: I believe the problem with `isUnitType` is that it doesn't consider enum literal types.

https://github.com/microsoft/TypeScript/pull/50423/files#r954112029

So we won't catch assignability from an enum literal to a weak type with no overlap.

```ts
enum E { A = "A" }

// these will differ
let x: { nope?: any } = E.A;
let y: { nope?: any } = "A";
```

Comment (ahejlsberg) [reply src/compiler/checker.ts:19113]: Enum literal types are definitely considered unit types. It's a bit subtle, `isUnitType` checks for `TypeFlags.Unit`, which includes `TypeFlags.Literal`, which includes `TypeFlags.StringLiteral` and `TypeFlags.NumberLiteral`. For enum literal types, the `TypeFlags.EnumLiteral` flag is always combined with one of the latter two, so `isUnitType` will always be `true`.

Comment (ahejlsberg) [reply src/compiler/checker.ts:19113]: Just to be sure, I verified both of your examples generate errors.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:19113]: Great! Can you add that as a test case?

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**